sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/pkversion/PartitionSupervisorImpl.java

```diff
@@ -61,11 +69,27 @@ public Mono<Void> run(CancellationToken shutdownToken) {
 
         return Mono.just(this)
             .delayElement(Duration.ofMillis(100), CosmosSchedulers.COSMOS_PARALLEL)
-            .repeat( () -> !shutdownToken.isCancellationRequested() && this.processor.getResultException() == null && this.renewer.getResultException() == null)
+            .repeat( () -> shouldContinue(shutdownToken))
             .last()
             .flatMap( value -> this.afterRun(context, shutdownToken));
     }
 
+    private boolean shouldContinue(CancellationToken shutdownToken) {
+        Duration timeSinceLastVerification = Duration.between(this.lastVerification, Instant.now());
+        if (timeSinceLastVerification.getSeconds() > this.renewer.getLeaseRenewInterval().getSeconds() * VERIFICATION_FACTOR) {
+            // if cfp has seen successes processing, we do a renew,
+            // otherwise we do not to allow lease stealing
+            if (processor.getProcessedBatches()) {
```

Comment:
hmm, I think in the processor, only have a boolean ProcessedBatches is not enough - there are cases where the whole flux pipeline can get stuck, then ProcessedBatches will be true forever

I think what we need is the last processed timestamp -> and then if the duration > this.renewer.getLeaseRenewInterval().getSeconds() * VERIFICATION_FACTOR, then just cancel both